samza-core/src/main/java/org/apache/samza/operators/impl/WindowOperatorImpl.java

```diff
@@ -18,26 +18,323 @@
+    }
+
+    WindowOperatorImpl.super.propagateResult((WM) paneOutput, collector, coordinator);
+  }
+
+  private void cancelTrigger(TriggerKey<K> triggerKey) {
+    TriggerImplWrapper wrapper = triggers.remove(triggerKey);
+    if (wrapper != null) {
+      wrapper.impl.cancel();
+    }
+  }
+
+  /**
+   * Implementation of the {@link TriggerContext} that allows {@link TriggerImpl}s to schedule and cancel callbacks.
+   */
+  public class TriggerContextImpl implements TriggerContext {
+    private final TriggerKey<K> triggerKey;
+
+    public TriggerContextImpl(TriggerKey<K> triggerKey) {
+      this.triggerKey = triggerKey;
+    }
+
+    @Override
+    public Cancellable scheduleCallback(Runnable runnable, long callbackTimeMs) {
+      TriggerTimerState timerState = new TriggerTimerState(triggerKey, runnable, callbackTimeMs);
+      pendingCallbacks.add(timerState);
+      return timerState;
+    }
+  }
+
+  private static class TriggerHandlerImpl implements TriggerImpl.TriggerCallbackHandler {
+
+    private boolean triggered;
+
+    @Override
+    public void onTrigger() {
+      triggered = true;
+    }
+
+    public boolean isTriggered() {
+      return triggered;
+    }
+
+    public void clearTrigger() {
+      triggered = false;
+    }
+  }
+
+  private class TriggerTimerState implements Comparable<TriggerTimerState>, Cancellable {
+
+    private final TriggerKey<K> triggerKey;
+    private final Runnable callback;
+
+    //the time in milliseconds at which the callback should trigger
+    private final long scheduleTimeMs;
+
+    private TriggerTimerState(TriggerKey<K> triggerKey, Runnable callback, long scheduleTimeMs) {
+      this.triggerKey = triggerKey;
+      this.callback = callback;
+      this.scheduleTimeMs = scheduleTimeMs;
+    }
+
+    private Runnable getCallback() {
+      return callback;
+    }
+
+    private long getScheduleTimeMs() {
+      return scheduleTimeMs;
+    }
+
+    @Override
+    public int compareTo(TriggerTimerState other) {
+      return Long.compare(this.scheduleTimeMs, other.scheduleTimeMs);
+    }
+
+    @Override
+    public boolean cancel() {
+      return pendingCallbacks.remove(this);
+    }
+  }
+
+  private class TriggerImplWrapper {
+    private final TriggerKey<K> triggerKey;
+    private final TriggerImpl<M> impl;
+    private final TriggerContext context;
+    private final TriggerHandlerImpl handler;
+
+    public TriggerImplWrapper(TriggerKey<K> triggerKey, TriggerImpl<M> impl, TriggerContext context, TriggerHandlerImpl handler) {
+      this.triggerKey = triggerKey;
+      this.impl = impl;
+      this.context = context;
+      this.handler = handler;
+    }
+
+    public void onMessage(M message, MessageCollector collector, TaskCoordinator coordinator) {
+      impl.onMessage(message, context, handler);
+      if (handler.isTriggered()) {
+        //repeating trigger can trigger multiple times, So, clear the handler to allow future triggerings.
+        handler.clearTrigger();
```

Comment:
Yup, you could totally do that and it would be equivalent. However,

1. It would push the complexity of managing the flags to the `TriggerImpl` classes.
2. It would add quite a bit of boiler plate to each of the `TriggerImpl` implementation classes.

This approach makes the `TriggerImpl`s much simpler (maybe useful when we allow for custom trigger impls later.)